libs/wire-subsystems/src/Wire/UserSubsystem/Interpreter.hs

```diff
@@ -495,3 +493,67 @@ checkHandlesImpl check num = reverse <$> collectFree [] check num
           case owner of
             Nothing -> collectFree (h : free) hs (n - 1)
             Just _ -> collectFree free hs n
+
+--------------------------------------------------------------------------------
+-- getting user accounts by different criteria
+
+getAccountsByImpl ::
+  forall r.
+  ( Member UserStore r,
+    Member DeleteQueue r,
+    Member (Input UserSubsystemConfig) r,
+    Member InvitationCodeStore r,
+    Member UserKeyStore r,
+    Member ActivationCodeStore r,
+    Member TinyLog r
+  ) =>
+  Local GetBy ->
+  Sem r [UserAccount]
+getAccountsByImpl (tSplit -> (domain, MkGetBy {includePendingInvitations, getByEmail, getByHandle, getByUserIds})) = do
+  config <- input
+
+  handleUserIds <- wither lookupHandle getByHandle
+
+  let storedToAcc = mkAccountFromStored domain config.defaultLocale
+
+      accountValid :: StoredUser -> Sem r (Maybe UserAccount)
+      accountValid storedUser =
+        let account = storedToAcc storedUser
+            notValid = pure Nothing
+            valid = pure $ Just account
+         in case userIdentity . accountUser $ account of
+              Nothing -> notValid
+              Just ident -> case (accountStatus account, includePendingInvitations, emailIdentity ident) of
+                (PendingInvitation, False, _) -> notValid
+                (PendingInvitation, True, Just email) ->
+                  lookupInvitationByEmail email >>= \case
+                    Nothing -> do
+                      -- user invited via scim should expire together with its invitation
+                      -- FIXME(mangoiv): this is not the right place to do this, ideally this should be part of a recurring
+                      -- job akin to 'pendingUserActivationCleanup'
+                      enqueueUserDeletion (userId account.accountUser)
+                      notValid
+                    Just _ -> valid
+                (PendingInvitation, True, Nothing) -> valid -- cannot happen, user invited via scim always has an email
+                (Active, _, _) -> valid
+                (Suspended, _, _) -> valid
+                (Deleted, _, _) -> valid
+                (Ephemeral, _, _) -> valid
+
+  accsByIds :: [UserAccount] <-
+    wither accountValid
+      =<< getUsers (nubOrd $ handleUserIds <> getByUserIds)
+
+  accsByEmail <- flip foldMap getByEmail \email -> do
+    let ek = mkEmailKey email
+    mactiveUid <- lookupKey ek
+    ac <- lookupActivationCode ek
+    let muidFromActivationKey = ac >>= fst
+    res <- getUsers (nubOrd $ catMaybes [mactiveUid, muidFromActivationKey])
+    pure $
+      map
+        storedToAcc
+        if includePendingInvitations
+          then res
+          else filter (\acc -> acc.status /= Just PendingInvitation) res
+  pure (nubOrd $ accsByIds <> accsByEmail)
```

Comment:
all of these should ideally be `Set` - but I don't know if it's a good idea to add a `FUTUREWORK`...